File: Grayjay.ClientServer/Models/ExceptionModel.cs
using Grayjay.Engine.Exceptions;

namespace Grayjay.ClientServer.Models
{
    public class ExceptionModel
    {
        public string Type { get; set; } = EXCEPTION_GENERAL;
        public string TypeName { get; set; }
        public string? Title { get; set; }
        public string? Message { get; set; }
        public string? Code { get; set; }
        public string? Stacktrace { get; set; }

        public string? PluginID { get; set; }
        public string? PluginName { get; set; }

        public bool CanRetry { get; set; } = false;


        public static ExceptionModel FromException(Exception ex, bool canRetry = false) => FromException(null, ex, canRetry);
        public static ExceptionModel FromException(string title, Exception ex, bool canRetry = false)
        {
            if (ex is ScriptException scriptEx)
            {
                string plugin = scriptEx.Config.Name;
                return new ExceptionModel()
                {
                    Type = EXCEPTION_SCRIPT,
                    Title = (!string.IsNullOrEmpty(title)) ? $"[{scriptEx.Config.Name}] {title}" : $"[{scriptEx.Config.Name}] {ex.Message}",
                    Message = (!string.IsNullOrEmpty(title)) ? ex.Message : null,
                    Code = scriptEx.Code,
                    CanRetry = canRetry,
                    TypeName = ex?.GetType()?.Name,
                    PluginID = scriptEx.Config.ID,
                    PluginName = scriptEx.Config.Name
                };
            }
            return new ExceptionModel()
            {
                Type = EXCEPTION_GENERAL,
                Title = (!string.IsNullOrEmpty(title)) ? title : ex.Message,
                Message = (!string.IsNullOrEmpty(title)) ? ex.Message : null,
                CanRetry = canRetry,
                TypeName = ex?.GetType()?.Name,
                Stacktrace = ex?.StackTrace
            };
        }


        public const string EXCEPTION_GENERAL = "general";
        public const string EXCEPTION_SCRIPT = "script";
    }
}

File: Grayjay.Desktop.Web/src/backend/exceptions/ExceptionModel.ts


export interface IExceptionModel {
    type: string,
    typeName?: string;
    title: string,
    message: string,
    code: string
    canRetry: boolean
    pluginID?: string;
    pluginName?: string;
    stacktrace?: string;

    model?: IExceptionModel;
}

export default class ExceptionModel extends Error implements IExceptionModel {
    type: string;
    typeName?: string;
    title: string;
    message: string;
    code: string;
    canRetry: boolean;

    pluginID?: string;
    pluginName?: string;

    constructor(model: IExceptionModel) {
        super();
        this.type = model.type ?? model?.model?.type;
        this.typeName = model.typeName ?? model?.model?.typeName;
        this.title =  model.title ?? model?.model?.title;
        this.message = model.message ?? model?.model?.message;
        this.code = model.code ?? model?.model?.code;
        this.canRetry = model.canRetry ?? model?.model?.canRetry;
        this.pluginID = model.pluginID ?? model?.model?.pluginID;
        this.pluginName = model.pluginName ?? model?.model?.pluginName;
        if(!this.code && model.stacktrace)
            this.code = model.stacktrace;
    }

    replaceTitle = (newTitle: string) => {
        if(!this.message)
            this.message = this.title;
        this.title = newTitle;
        return this;
    }
}